group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilderTest.java

```diff
@@ -0,0 +1,1070 @@
+            topic1Uuid,
+            topic1Name,
+            5,
+            mkMapOfPartitionRacks(3)
+        ));
+        topicMetadata.put(topic2Uuid, new TopicMetadata(
+            topic2Uuid,
+            topic2Name,
+            2,
+            mkMapOfPartitionRacks(3)
+        ));
+
+        // Initial subscriptions were [T1, T2].
+        Map<String, AssignmentMemberSpec> members = new TreeMap<>();
+
+        Map<Uuid, Set<Integer>> currentAssignmentForA = new TreeMap<>(
+            mkAssignment(
+                mkTopicAssignment(topic1Uuid, 0, 2, 4),
+                mkTopicAssignment(topic2Uuid, 0)
+            )
+        );
+        members.put(memberA, new AssignmentMemberSpec(
+            Optional.empty(),
+            Optional.of("rack0"),
+            Collections.singletonList(topic1Uuid),
+            currentAssignmentForA
+        ));
+
+        Map<Uuid, Set<Integer>> currentAssignmentForB = new TreeMap<>(
+            mkAssignment(
+                mkTopicAssignment(topic1Uuid, 1, 3),
+                mkTopicAssignment(topic2Uuid, 1)
+            )
+        );
+        members.put(memberB, new AssignmentMemberSpec(
+            Optional.empty(),
+            Optional.of("rack1"),
+            Collections.singletonList(topic1Uuid),
+            currentAssignmentForB
+        ));
+
+        AssignmentSpec assignmentSpec = new AssignmentSpec(members);
+        SubscribedTopicMetadata subscribedTopicMetadata = new SubscribedTopicMetadata(topicMetadata);
+
+        GroupAssignment computedAssignment = assignor.assign(assignmentSpec, subscribedTopicMetadata);
+
+        Map<String, Map<Uuid, Set<Integer>>> expectedAssignment = new HashMap<>();
+        expectedAssignment.put(memberA, mkAssignment(
+            mkTopicAssignment(topic1Uuid, 0, 2, 4)
+        ));
+        expectedAssignment.put(memberB, mkAssignment(
+            mkTopicAssignment(topic1Uuid, 1, 3)
+        ));
+
+        assertAssignment(expectedAssignment, computedAssignment);
+        checkValidityAndBalance(members, computedAssignment);
+    }
+
+    /**
+     * Verifies that the given assignment is valid with respect to the given subscriptions.
+     * Validity requirements:
+     *      - each member is subscribed to topics of all partitions assigned to it, and
+     *      - each partition is assigned to no more than one member.
+     * Balance requirements:
+     *      - the assignment is fully balanced (the numbers of topic partitions assigned to members differ by at most one), or
+     *      - there is no topic partition that can be moved from one member to another with 2+ fewer topic partitions.
+     *
+     * @param members                   members data structure from assignmentSpec
+     * @param computedGroupAssignment   given assignment for balance check
+     */
+    private void checkValidityAndBalance(Map<String, AssignmentMemberSpec> members, GroupAssignment computedGroupAssignment) {
```

Comment:
Well since we check the exact assignment for each member, when I was writing the exact partitions I made sure that the racks are matching so I feel like that is checked there already. To write a check again we would have to do quite some steps since its not guaranteed that every partition assigned to a member has a matching rack even if both partition and member racks exist

Reply:
that makes sense

Reply:
can we put the arguments into new lines?